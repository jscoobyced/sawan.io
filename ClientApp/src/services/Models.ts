export enum Language {
    English = 0,
    French = 1
}

export interface Link {
    text: string;
    url: string;
    target?: string;
    title?: string;
}

export interface MainContent {
    navigationMenuContent: NavigationMenuContent;
    menuContent: MenuContent;
    footerContent: FooterContent;
}

export interface NavigationMenuContent {
    home: string;
    websiteName: string;
    applications: string;
    cryptoCurrency: string;
    healthMonitor: string;
    about: string;
    information: string;
    resume: string;
}

export interface BlogElement {
    id: number;
    blogDate: Date;
    article: string;
    articleTitle: string;
}

export interface BlogPage {
    articles: BlogElement[];
}

export interface MenuContent {
    links: Link[];
}

export interface FooterContent {
    copyright: string;
    credits: string;
    year: string;
}
