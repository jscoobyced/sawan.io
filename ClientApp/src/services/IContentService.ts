import { BlogElement, BlogPage, Language, MainContent } from './Models';

export interface IContentService {
    getMainContent(language: Language): Promise<MainContent>;
    getDefaultMainContent(): MainContent;
    getBlogPage(maxResult: number): Promise<BlogPage>;
    getDefaultBlogPage(): BlogPage;
    getBlogElement(id: number): Promise<BlogElement>;
}
