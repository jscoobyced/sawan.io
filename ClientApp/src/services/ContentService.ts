import { ApiUtils } from '../utils/ApiUtils';
import { IContentService } from './IContentService';
import { BlogElement, BlogPage, Language, MainContent } from './Models';

export class ContentService implements IContentService {

    public async getMainContent(language: Language): Promise<MainContent> {
        return this.fetchMainContent(language);
    }

    public getDefaultMainContent(): MainContent {
        return this.defaultMainContent();
    }

    public async getBlogPage(yearMonth: string, maxResult: number): Promise<BlogPage> {
        return this.updateBlogPage(yearMonth, maxResult);
    }

    public async getBlogElement(id: number): Promise<BlogElement> {
        return ApiUtils.fetchData<BlogElement>(`/api/Data/blog/${id}`);
    }

    public getDefaultBlogPage(): BlogPage {
        return {
            articles: []
        };
    }

    public async saveBlogElement(blogElement: BlogElement): Promise<boolean> {
        return this.doSaveBlogElement(blogElement);
    }

    protected async doSaveBlogElement(blogElement: BlogElement): Promise<boolean> {
        return ApiUtils.postData('/api/Data/blog/post/', blogElement);
    }

    protected async fetchMainContent(language: Language): Promise<MainContent> {
        return ApiUtils.fetchData<MainContent>(`/api/Data/main/${language}`);
    }

    protected defaultMainContent(): MainContent {
        return {
            menuContent: {
                title: '',
                links: []
            },
            navigationMenuContent: {
                about: '',
                applications: '',
                cryptoCurrency: '',
                healthMonitor: '',
                home: '',
                websiteName: '',
                information: '',
                resume: ''
            },
            footerContent: {
                copyright: '',
                credits: '',
                year: ''
            }
        };
    }

    protected async updateBlogPage(yearMonth: string, maxResult: number): Promise<BlogPage> {
        const empty = {
            articles: []
        };
        if (maxResult <= 0) {
            return empty;
        }
        const data = await ApiUtils.fetchData<BlogElement[]>(`/api/Data/blogpage/${yearMonth}/${maxResult}`);
        if (!data || data.length === 0) {
            return empty;
        }
        return {
            articles: data.slice(0, Math.min(maxResult, data.length))
        };
    }
}
