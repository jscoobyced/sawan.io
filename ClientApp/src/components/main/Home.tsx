import React from 'react';
import { BlogElement, BlogPage } from '../../services/Models';
import { AuthenticationFactory } from '../../utils/AuthenticationFactory';
import { Article } from '../blog/Article';

export interface HomeProps {
    blogPage: BlogPage;
    isSignedIn: boolean;
}

export interface HomeState {
    isSignedIn: boolean;
}

export class Home extends React.Component<HomeProps, HomeState> {

    public constructor(props: HomeProps, state: HomeState) {
        super(props, state);
        this.state = {
            isSignedIn: false
        };
    }

    public render() {
        const articles: BlogElement[] = this.props.blogPage.articles;
        const isAdmin = AuthenticationFactory.getAuthentication() && AuthenticationFactory.getAuthentication().isAdmin();

        if (!articles || articles.length === 0) {
            return null;
        }

        return articles.map((article, key) => {
            return (
                <Article
                    key={key}
                    blogElement={article}
                    backLink={false}
                    editLink={isAdmin}
                    ellipsis={true}
                />);
        });
    }
}
