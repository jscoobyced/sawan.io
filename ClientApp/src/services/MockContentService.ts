import { DateUtil } from '../utils/DateUtils';
import { ContentService } from './ContentService';
import { BlogElement, BlogPage, Language, MainContent } from './Models';

export class MockContentService extends ContentService {

    private readonly data = 'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do'
        + ' ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco'
        + ' aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit '
        + ' cillum dolore eu fugiat nulla pariatur.<br />Excepteur sint occaecat cupidatat non proident, '
        + ' culpa qui officia deserunt mollit anim id est laborum. Commodo qui imperdiet massa tincidunt.'
        + ' Non enim praesent elementum facilisis leo. Adipiscing diam donec adipiscing tristique. Velit'
        + ' euismod in pellentesque massa placerat. Interdum velit euismod in pellentesque. Phasellus'
        + ' vestibulum lorem sed risus ultricies tristique nulla. Egestas erat imperdiet sed euismod nisi'
        + ' porta lorem mollis aliquam. Nulla facilisi etiam dignissim diam quis enim lobortis scelerisque'
        + ' fermentum. Morbi tristique senectus et netus et malesuada fames.<br />Vitae suscipit tellus'
        + ' mauris a diam maecenas. Sed risus ultricies tristique nulla aliquet. Sagittis eu volutpat odio'
        + ' facilisis mauris sit amet. Tempor orci eu lobortis elementum nibh tellus molestie nunc non.'
        + ' Turpis nunc eget lorem dolor.';
    private readonly displayDate = DateUtil.defaultDate();
    private readonly articles: BlogElement[] = [
        {
            id: '1',
            blogDate: this.displayDate,
            articleTitle: 'Starting again',
            article: 'A new begining this year for me. After 5 years at Agoda it is time for me to move on to'
                + ' new challenges. I am going to do something I haven\'t done for a while: writing a '
                + 'resume that reflects my skills and experience without showing off. Not too much at least '
                + '<i class=\'far fa-smile\'></i>'
                + '<br />So I\'m going to take this opportunity to create it using good CSS3 and HTML5 practices.'
                + ' The first thing I have in mind is to limit the DOM depth. Resumes have simple structures, so'
                + ' we shouldn\'t need a long list of nested DIVs. Second is to chosing a monospaced font so we can'
                + ' keep control of the format when printing. Let\'s throw in some '
                + '<a target=\'_blank\' href=\'https://fontawesome.com/\'>font awesome</a> to have a nice touch on '
                + 'the contact details.<br /> You can see the current version <a href=\'/resume/?full\' target=\'_blank\'>'
                + 'here</a>.'
        },
        {
            id: '2',
            blogDate: this.displayDate,
            articleTitle: 'Grand opening of sawan.io',
            article: 'Grand opening of sawan.io.<br />This website is about my'
                + ' personal training on various web-technologies: TypeScript, webpack, ReactJS, Secure HTTP'
                + '  headers... It includes a simple (and a bit naive) crypto-currency analyser: it can'
                + ' suggest you to sell or buy. Use at your own risk, it is amateur predictions, you have'
                + ' been warned.<br />There will be also an application to track health metrics (weight,'
                + ' % body fat, vascular fat ratio...). Or maybe not, this might be my next project.<br />'
                + ' This project is open-sourced on <a target=\'_blank\''
                + ' href=\'https://github.com/jscoobyced/sawan.io\'>Github</a>. Enjoy and have fun.'
        },
        {
            id: '3',
            blogDate: this.displayDate,
            articleTitle: 'Some other news',
            article: this.data
        }
    ];

    public async getBlogElement(id: string): Promise<BlogElement> {
        return this.updateBlogPage(3).then(blogPage => {
            let article = null as unknown as BlogElement;
            for (const a of blogPage.articles) {
                if (a.id === id) {
                    article = a;
                    break;
                }
            }

            return article;
        });
    }

    protected updateMainContent(language: Language): Promise<MainContent> {
        const content: MainContent = {
            menuContent: {
                links: []
            },
            navigationMenuContent: {
                about: 'About',
                applications: 'Applications',
                cryptoCurrency: 'Crypto-Currency',
                healthMonitor: 'Health Monitor',
                home: 'Home',
                websiteName: 'sawan.io',
                information: 'Information',
                resume: 'Resume'
            },
            footerContent: {
                copyright: 'Copyright',
                credits: 'sawan.io',
                year: '2018 - ' + new Date().getFullYear()
            }
        };
        return Promise.resolve(content);
    }

    protected async doSaveBlogElement(blogElement: BlogElement): Promise<boolean> {
        for (const article of this.articles) {
            if (article.id === blogElement.id) {
                article.article = blogElement.article;
                article.articleTitle = blogElement.articleTitle;
            }
        }
        return Promise.resolve(true);
    }

    protected updateBlogPage = (maxResult: number): Promise<BlogPage> => {
        const blog: BlogPage = {
            articles: this.articles
        };
        return Promise.resolve(blog);
    }
}
