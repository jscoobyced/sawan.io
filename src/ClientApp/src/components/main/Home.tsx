import React from 'react';
import { NavLink } from 'react-router-dom';
import xss from 'xss';
import { HomeContent } from '../../services/Models';

export interface HomeProps {
    homeContent: HomeContent;
}

export class Home extends React.Component<HomeProps> {

    private readonly options = {
        whiteList: {
            a: ['href', 'title', 'target'],
            br: []
        }
    };
    private readonly xssFilter = new xss.FilterXSS(this.options);

    public constructor(props: HomeProps) {
        super(props);
    }

    public render() {
        const articles: string[] = this.props.homeContent.article as string[];
        const titles: string[] = this.props.homeContent.articleTitle as string[];
        if (!articles || !titles || articles.length !== titles.length) {
            return <span></span>;
        }

        return articles.map((article, key) => <article key={key}>
            <h1>{titles[key]}</h1>
            <div>
                <span dangerouslySetInnerHTML={{ __html: this.getSafeContent(article) }}></span>
                <NavLink to='/' className='link-article'>Read more...</NavLink>
            </div>
        </article>);
    }

    private getSafeContent(content: string): string {
        const safeContent = this.xssFilter.process(content);
        let safeSummary = safeContent.substring(0, 1000);
        if (safeContent !== safeSummary) {
            safeSummary = safeSummary.substring(0, safeSummary.lastIndexOf(" "));
            safeSummary += '&#8230;';
        }
        return safeSummary;
    }
}
