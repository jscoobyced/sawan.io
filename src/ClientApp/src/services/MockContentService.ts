import { ContentService } from './ContentService';
import { BlogElement, BlogPage, Language, MainContent } from './Models';

export class MockContentService extends ContentService {

    private readonly data = 'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do'
        + ' ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco'
        + ' aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit '
        + ' cillum dolore eu fugiat nulla pariatur.<br />Excepteur sint occaecat cupidatat non proident, '
        + ' culpa qui officia deserunt mollit anim id est laborum. Commodo qui imperdiet massa tincidunt.'
        + ' Non enim praesent elementum facilisis leo. Adipiscing diam donec adipiscing tristique. Velit'
        + ' euismod in pellentesque massa placerat. Interdum velit euismod in pellentesque. Phasellus'
        + ' vestibulum lorem sed risus ultricies tristique nulla. Egestas erat imperdiet sed euismod nisi'
        + ' porta lorem mollis aliquam. Nulla facilisi etiam dignissim diam quis enim lobortis scelerisque'
        + ' fermentum. Morbi tristique senectus et netus et malesuada fames.<br />Vitae suscipit tellus'
        + ' mauris a diam maecenas. Sed risus ultricies tristique nulla aliquet. Sagittis eu volutpat odio'
        + ' facilisis mauris sit amet. Tempor orci eu lobortis elementum nibh tellus molestie nunc non.'
        + ' Turpis nunc eget lorem dolor.';

    protected updateMainContent(language: Language): Promise<MainContent> {
        const content: MainContent = {
            menuContent: {
                links: []
            },
            navigationMenuContent: {
                about: 'About',
                applications: 'Applications',
                cryptoCurrency: 'Crypto-Currency',
                healthMonitor: 'Health Monitor',
                home: 'Home',
                websiteName: 'sawan.io'
            },
            footerContent: {
                copyright: 'Copyright',
                credits: 'sawan.io',
                year: '2018 - ' + new Date().getFullYear()
            }
        };
        return Promise.resolve(content);
    }

    protected updateBlogPage(maxResult: number): Promise<BlogPage> {
        const date = '2018-11-15';
        const displayDate = (new Date(date).toISOString().substring(0, 10));
        const articles: BlogElement[] = [
            {
                article: `Latest news - ${displayDate}`,
                articleTitle: 'Grand opening of sawan.io.<br />This website is about my'
                    + ' personal training on various web-technologies: TypeScript, webpack, ReactJS, Secure HTTP'
                    + '  headers... It includes a simple (and a bit naive) crypto-currency analyser: it can'
                    + ' suggest you to sell or buy. Use at your own risk, it is amateur predictions, you have'
                    + ' been warned.<br />There will be also an application to track health metrics (weight,'
                    + ' % body fat, vascular fat ratio...). Or maybe not, this might be my next project.<br />'
                    + ' This project is open-sourced on <atarget=\'_blank\''
                    + ' href=\'https://github.com/jscoobyced/sawan.io\'>Github</a>. Enjoy and have fun.'
            },
            {
                article: 'Lorem Ipsum',
                articleTitle: this.data
            },
            {
                article: 'Some other news',
                articleTitle: this.data
            }
        ];
        return Promise.resolve({ articles });
    }
}
