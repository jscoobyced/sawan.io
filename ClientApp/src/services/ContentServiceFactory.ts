import { ContentService } from './ContentService';
import { IContentService } from './IContentService';
import { MockContentService } from './MockContentService';

export class ContentServiceFactory {

    public static GetContentService(): IContentService {
        const mode = process.env.mode as string;
        if (mode === 'production') {
            return new ContentService();
        }

        return new MockContentService();
    }
}
