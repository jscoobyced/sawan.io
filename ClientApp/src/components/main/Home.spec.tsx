import { shallow } from 'enzyme';
import React from 'react';
import { MockContentService } from '../../services/MockContentService';
import { Home } from './Home';

test('Home component is unchanged.', async () => {
    const blogPage = await new MockContentService().getBlogPage(4);
    const home = shallow(<Home blogPage={blogPage} />);
    expect(home).toMatchSnapshot();
});

test('Home component with no article.', () => {
    const blogPage = new MockContentService().getDefaultBlogPage();
    const home = shallow(<Home blogPage={blogPage} />);
    expect(home).toMatchSnapshot();
});

test('Home component signed-in.', () => {
    const blogPage = new MockContentService().getDefaultBlogPage();
    const home = shallow(<Home blogPage={blogPage} />);
    expect(home).toMatchSnapshot();
});

