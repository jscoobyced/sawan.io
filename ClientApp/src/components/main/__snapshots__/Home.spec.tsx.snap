// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home component is unchanged. 1`] = `
Array [
  <Article
    backLink={false}
    blogElement={
      Object {
        "article": "A new begining this year for me. After 5 years at Agoda it is time for me to move on to new challenges. I am going to do something I haven't done for a while: writing a resume that reflects my skills and experience without showing off. Not too much at least <i class='far fa-smile'></i><br />So I'm going to take this opportunity to create it using good CSS3 and HTML5 practices. The first thing I have in mind is to limit the DOM depth. Resumes have simple structures, so we shouldn't need a long list of nested DIVs. Second is to chosing a monospaced font so we can keep control of the format when printing. Let's throw in some <a target='_blank' href='https://fontawesome.com/'>font awesome</a> to have a nice touch on the contact details.<br /> You can see the current version <a href='/resume/?full' target='_blank'>here</a>.",
        "articleTitle": "Starting again",
        "blogDate": 2019-01-01T00:00:00.000Z,
        "id": 1,
        "updateDate": 2019-01-01T00:00:00.000Z,
      }
    }
    ellipsis={true}
    key="0"
  />,
  <Article
    backLink={false}
    blogElement={
      Object {
        "article": "Grand opening of sawan.io.<br />This website is about my personal training on various web-technologies: TypeScript, webpack, ReactJS, Secure HTTP  headers... It includes a simple (and a bit naive) crypto-currency analyser: it can suggest you to sell or buy. Use at your own risk, it is amateur predictions, you have been warned.<br />There will be also an application to track health metrics (weight, % body fat, vascular fat ratio...). Or maybe not, this might be my next project.<br /> This project is open-sourced on <a target='_blank' href='https://github.com/jscoobyced/sawan.io'>Github</a>. Enjoy and have fun.",
        "articleTitle": "Grand opening of sawan.io",
        "blogDate": 2019-01-01T00:00:00.000Z,
        "id": 2,
        "updateDate": 2019-01-01T00:00:00.000Z,
      }
    }
    ellipsis={true}
    key="1"
  />,
  <Article
    backLink={false}
    blogElement={
      Object {
        "article": "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit  cillum dolore eu fugiat nulla pariatur.<br />Excepteur sint occaecat cupidatat non proident,  culpa qui officia deserunt mollit anim id est laborum. Commodo qui imperdiet massa tincidunt. Non enim praesent elementum facilisis leo. Adipiscing diam donec adipiscing tristique. Velit euismod in pellentesque massa placerat. Interdum velit euismod in pellentesque. Phasellus vestibulum lorem sed risus ultricies tristique nulla. Egestas erat imperdiet sed euismod nisi porta lorem mollis aliquam. Nulla facilisi etiam dignissim diam quis enim lobortis scelerisque fermentum. Morbi tristique senectus et netus et malesuada fames.<br />Vitae suscipit tellus mauris a diam maecenas. Sed risus ultricies tristique nulla aliquet. Sagittis eu volutpat odio facilisis mauris sit amet. Tempor orci eu lobortis elementum nibh tellus molestie nunc non. Turpis nunc eget lorem dolor.",
        "articleTitle": "Some other news",
        "blogDate": 2019-01-01T00:00:00.000Z,
        "id": 3,
        "updateDate": 2019-01-01T00:00:00.000Z,
      }
    }
    ellipsis={true}
    key="2"
  />,
]
`;

exports[`Home component signed-in. 1`] = `""`;

exports[`Home component with no article. 1`] = `""`;
